zebra-chain/src/transaction/txidhash.rs

```diff
@@ -0,0 +1,54 @@
+//! Transaction ID hashing. Contains code for generating the Transaction ID
+//! from the transaction, using hashing.
+use std::{convert::TryInto, io};
+
+use super::Transaction;
+
+use crate::serialization::{sha256d, ZcashSerialize};
+
+use super::Hash;
+
+/// A Transaction ID hasher. It computes the transaction ID by hashing
+/// different parts of the transaction, depending on the transaction version.
+/// For V5 transactions, it follows ZIP-244 and ZIP-225.
+pub(super) struct TxIdHasher<'a> {
+    trans: &'a Transaction,
+}
+
+impl<'a> TxIdHasher<'a> {
+    /// Return a new TxIdHasher for the given transaction.
+    pub fn new(trans: &'a Transaction) -> Self {
+        TxIdHasher { trans }
+    }
+
+    /// Compute the Transaction ID for the previously specified transaction.
+    pub(super) fn txid(self) -> Result<Hash, io::Error> {
+        match self.trans {
+            Transaction::V1 { .. }
+            | Transaction::V2 { .. }
+            | Transaction::V3 { .. }
+            | Transaction::V4 { .. } => self.txid_v1_to_v4(),
+            Transaction::V5 { .. } => self.txid_v5(),
+        }
+    }
+
+    /// Compute the Transaction ID for transactions V1 to V4.
+    /// In these cases it's simply the hash of the serialized transaction.
+    fn txid_v1_to_v4(self) -> Result<Hash, io::Error> {
+        let mut hash_writer = sha256d::Writer::default();
+        self.trans
+            .zcash_serialize(&mut hash_writer)
+            .expect("Transactions must serialize into the hash.");
+        Ok(Hash(hash_writer.finish()))
+    }
+
+    /// Compute the Transaction ID for a V5 transaction in the given network upgrade.
+    /// In this case it's the hash of a tree of hashes of specific parts of the
+    /// transaction, as specified in ZIP-244 and ZIP-225.
+    fn txid_v5(self) -> Result<Hash, io::Error> {
+        // The v5 txid (from ZIP-244) is computed using librustzcash. Convert the zebra
+        // transaction to a librustzcash transaction.
+        let alt_tx: zcash_primitives::transaction::Transaction = self.trans.try_into()?;
+        Ok(Hash(*alt_tx.txid().as_ref()))
+    }
+}
```

Comment:
PIggy-backing on what @jvff said, I think it'd be nicer if we impl'd `Transaction::id()` and all this logic lived in there; having a separate struct named `Hasher` that doesn't implement `std::hash::Hasher` seems confusing

Reply:
I'd like to split this change into two parts, to keep the scope contained:

1. We can make changes to the `txidhash` module in this PR. I think the separate methods make the code more readable, but I don't mind either way. Maybe the struct would be less confusing if it was named `ZcashTxId{Builder,Generator,Calculator}`.

2. There is already a `Transaction::hash` method, so let's not cause confusion by adding a `Transaction::id` method as well. If we want to change `hash` to `id`, let's open a follow-up ticket.

Reply:
I based this design on `sighash.rs` which defines a `SigHasher` (which also does not implement `std::hash::Hasher`).

I agree that the names are a bit confusing. I've changed `txidhash.rs` into `txid.rs` and `TxIdHasher` into `TxIdBuilder` in 92378a6b. (I decided to not use the `Zcash` prefix because it sounded a bit odd, since there's no other kind of TxId)

In #2438 I've added a item to consider incorporating this back into `Transaction`. But if for whatever reason we decide to  actually implement the txid logic instead of calling librustzcash then it may be useful have it in a separate file since the implementation will become much larger.